File: Ruby Solution/sortingTheFile.rb

startTime = Time.now
#Defining the constants
isMergeSort = true
isFullSort = false
inputFileName = 'hs-set-10k.txt'
outputFileName = 'hs-set-10k-output.txt'

#Opening the file and creating a struct
Card = Struct.new(:name, :className, :rarity, :set, :type, :cost)
file = File.open(inputFileName, 'r')
data = file.read
eachLineStringArray =  data.split("\r\n")

#creating an array struct and storing all elements
storedCards = []
sortedCards = []
for i in eachLineStringArray
  eachItem = i.split("\t")
  storedCards << Card.new(eachItem[0],eachItem[1],eachItem[2],eachItem[3],eachItem[4],eachItem[5])
end

def mergeSortFull(array)
  # Recursive Output
  if array.size <= 1
    return array
  end

  #Dividing array into parts
  mid = array.size/2
  part_a = mergeSortFull(array[0...mid])
  part_b = mergeSortFull(array[mid..array.size])

  # Merging the array
  array = []
  offset_a = 0
  offset_b = 0

  while offset_a < part_a.size and offset_b < part_b.size
    a = part_a[offset_a]
    b = part_b[offset_b]

    if a.className < b.className
      array << a
      offset_a += 1
    elsif a.className == b.className
      if a.cost.to_i < b.cost.to_i
        array << a
        offset_a += 1
      elsif a.cost.to_i == b.cost.to_i
        if a.name < b.name
          array << a
          offset_a += 1
        else
          array << b
          offset_b += 1
        end
      else
        array << b
        offset_b += 1
      end
    else
      array << b
      offset_b += 1
    end
  end

  while offset_a < part_a.count
    array << part_a[offset_a]
    offset_a += 1
  end
  while offset_b < part_b.count
    array << part_b[offset_b]
    offset_b += 1
  end
  return array
end


def mergeSortFilter(array)
  # Recursive Output
  if array.size <= 1
    return array
  end
  #Dividing array into parts
  mid = array.size/2
  part_a = mergeSortFilter(array[0...mid])
  part_b = mergeSortFilter(array[mid..array.size])

  # Merging the array
  array = []
  offset_a = 0
  offset_b = 0

  while offset_a < part_a.size and offset_b < part_b.size
    a = part_a[offset_a]
    b = part_b[offset_b]

    if a.type <= b.type
      array << a
      offset_a += 1
    else
      array << b
      offset_b += 1
    end
  end

  while offset_a < part_a.count
    array << part_a[offset_a]
    offset_a += 1
  end
  while offset_b < part_b.count
    array << part_b[offset_b]
    offset_b += 1
  end
  return array
end


if (isMergeSort && isFullSort)
    sortedCards = mergeSortFull(storedCards)

elsif (isMergeSort && !isFullSort)
    sortedCards = mergeSortFilter(storedCards)

elsif (!isMergeSort && isFullSort)

  (1...storedCards.size).each { |t|
    counter = t-1
    i = t
    while counter >= 0
      if storedCards[i].className < storedCards[counter].className
        temp = storedCards[counter]
        storedCards[counter] = storedCards[i]
        storedCards[i] = temp
        i -=1
      elsif storedCards[i].className == storedCards[counter].className
        if storedCards[i].cost.to_i < storedCards[counter].cost.to_i
          temp = storedCards[counter]
          storedCards[counter] = storedCards[i]
          storedCards[i] = temp
          i -= 1
        elsif storedCards[i].cost.to_i == storedCards[counter].cost.to_i
          if storedCards[i].name < storedCards[counter].name
            temp = storedCards[counter]
            storedCards[counter] = storedCards[i]
            storedCards[i] = temp
            i -= 1
          end
        end
      end
      counter -= 1
    end
  }
  sortedCards = storedCards

else

  (1...storedCards.size).each { |t|
    counter = t-1
    i = t
    while counter >= 0
      if storedCards[i].type < storedCards[counter].type
        temp = storedCards[counter]
        storedCards[counter] = storedCards[i]
        storedCards[i] = temp
        i -=1
      end
      counter -= 1
    end
  }
  sortedCards = storedCards

end

# write to a file
File.open(outputFileName, 'w') { |file|
  text = ""
  for i in sortedCards
    text += (i.name + "\t" + i.className + "\t" + i.rarity + "\t" + i.set + "\t" +  i.type + "\t" + i.cost + "\r\n")
  end
  file.write(text)
}

endTime = Time.now

puts "Running time is #{(endTime - startTime)} secs"

File: Swift Solution/sortingTheFile.swift
import Foundation


// Input variables you can change those variables
var file = "hs-set-10k.txt"
var isMergeSort = true
var isFullSort = true
var outputFileName = "hs-set-10k-out.txt"
var sortedArray = Array<(name: String, className: String, rarity: String, set: String, type: String, cost: String)>()
// Struct for cards
var cards = Array<(name: String, className: String, rarity: String, set: String, type: String, cost: String)>()

// Opening a file inside the documentation directory
if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {

    let fileURL = dir.appendingPathComponent(file)

    //reading from the file
    do {
        let allInput = try String(contentsOf: fileURL, encoding: .utf8)
        let tempArrayAllLines = allInput.components(separatedBy: "\r\n")

        // Spliting separed lines with "\t" and add it to the tuple list
        var i: Int = 0
        while i < (tempArrayAllLines.count-1) {
            let elementsInLine = tempArrayAllLines[i].components(separatedBy: "\t")
            cards.append((name: elementsInLine[0], className: elementsInLine[1], rarity: elementsInLine[2], set: elementsInLine[3], type: elementsInLine[4], cost: elementsInLine[5]))
            i += 1;
        }
    }
    catch {
        print("You gotta put file named <\(file)> on this directory -> <\(dir)>")
        print(error)
    }
}


func mergeSortFull(_ array: [(name: String, className: String, rarity: String, set: String, type: String, cost: String)]) -> [(name: String, className: String, rarity: String, set: String, type: String, cost: String)]{
    // Breaking the recursive output
    if array.count <= 1 {
        return array
    }

    // Dividing array into parts
    let mid = array.count/2
    var part_a = mergeSortFull(Array(array[0..<mid]))
    var part_b = mergeSortFull(Array(array[mid..<array.count]))

    // Merging parts that we broke
    var array = Array<(name: String, className: String, rarity: String, set: String, type: String, cost: String)>()
    var offset_a = 0
    var offset_b = 0
    while offset_a < part_a.count && offset_b < part_b.count{
        let a = part_a[offset_a]
        let b = part_b[offset_b]

        if a.className < b.className{
            array.append(a)
            offset_a += 1
        }else if a.className == b.className {
            if Int(a.cost)! < Int(b.cost)! {
                array.append(a)
                offset_a += 1
            }else if Int(a.cost) == Int(b.cost){
                if a.name <= b.name {
                    array.append(a)
                    offset_a += 1
                }else {
                    array.append(b)
                    offset_b += 1
                }
            }else {
                array.append(b)
                offset_b += 1
            }
        }else {
            array.append(b)
            offset_b += 1
        }
    }

    while offset_a < part_a.count {
        array.append(part_a[offset_a])
        offset_a += 1
    }
    while offset_b < part_b.count {
        array.append(part_b[offset_b])
        offset_b += 1
    }

    return array
}

func mergeSortFilter(_ array: [(name: String, className: String, rarity: String, set: String, type: String, cost: String)]) -> [(name: String, className: String, rarity: String, set: String, type: String, cost: String)]{
    // Breaking the recursive output
    if array.count <= 1 {
        return array
    }

    // Dividing array into parts
    let mid = array.count/2
    var part_a = mergeSortFilter(Array(array[0..<mid]))
    var part_b = mergeSortFilter(Array(array[mid..<array.count]))

    // Merging parts that we broke
    var array = Array<(name: String, className: String, rarity: String, set: String, type: String, cost: String)>()
    var offset_a = 0
    var offset_b = 0
    while offset_a < part_a.count && offset_b < part_b.count{
        let a = part_a[offset_a]
        let b = part_b[offset_b]

        if a.type <= b.type {
            array.append(a)
            offset_a += 1
        }else {
            array.append(b)
            offset_b += 1
        }
    }

    while offset_a < part_a.count {
        array.append(part_a[offset_a])
        offset_a += 1
    }
    while offset_b < part_b.count {
        array.append(part_b[offset_b])
        offset_b += 1
    }

    return array
}

let programStart = Date()

if isFullSort && isMergeSort{
    // Full sort with merge
    sortedArray = mergeSortFull(cards)
}else if !isFullSort && isMergeSort{
    // Filter sort with merge
    sortedArray = mergeSortFilter(cards)
}else if !isFullSort && !isMergeSort{
    // Filter sort with insertion
    var counterOne = 0
    var keyPosition: Int
    var counterTwo: Int

    for key in cards {
        counterTwo = counterOne - 1
        keyPosition = counterOne
        while counterTwo >= 0 {
            // When key is smaller change the key with the other number
            if (cards[counterTwo].type > key.type){
                cards[keyPosition] = cards[counterTwo]
                keyPosition -= 1
            }
            counterTwo -= 1
        }
        cards[keyPosition] = key
        counterOne += 1
    }
    sortedArray = cards
}else{
    // Full sort with insertion
    var counterOne = 0
    var keyPosition: Int
    var counterTwo: Int

    for key in cards {
        counterTwo = counterOne - 1
        keyPosition = counterOne
        while counterTwo >= 0 {
            // When key is smaller change the key with the other number
            if (cards[counterTwo].className > key.className){
                cards[keyPosition] = cards[counterTwo]
                keyPosition -= 1
            }else if (cards[counterTwo].className == key.className) {
                if (Int(cards[counterTwo].cost)! > Int(key.cost)!){
                    cards[keyPosition] = cards[counterTwo]
                    keyPosition -= 1
                }else if (Int(cards[counterTwo].cost)! == Int(key.cost)!) {
                    if (cards[counterTwo].name > key.name){
                        cards[keyPosition] = cards[counterTwo]
                        keyPosition -= 1
                    }
                }
            }
            counterTwo -= 1
        }
        cards[keyPosition] = key
        counterOne += 1
    }
    sortedArray = cards
}
let programFinish = Date()


if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {

    let fileURL = dir.appendingPathComponent(outputFileName)

    //writing to the file
    do {
        var i = 0
        var text = ""
        while i < cards.count{
            text += sortedArray[i].name + "\t" + sortedArray[i].className + "\t" + sortedArray[i].rarity + "\t" + sortedArray[i].set + "\t" + sortedArray[i].type + "\t" + sortedArray[i].cost + "\r\n"

            i += 1
        }
        try text.write(to: fileURL, atomically: false, encoding: .utf8)
    }
    catch {
        print(error)
    }
}


let executionTime = programFinish.timeIntervalSince(programStart)
print("Execution time: \(executionTime), is Full sort type: \(isFullSort), is Merge sort: \(isMergeSort)")

